### ISSUE TITLE
Accuracy and resolution of sleep() on Linux should be improved

### ISSUE THREAD

**Issue by ufechner7** [JuliaLang/julia#12770]
On Linux, the resolution of the function sleep() could be much better. 
The following test script:
https://gist.github.com/ufechner7/d264fb714a551d333e6b
shows an error of about 1.2 ms on my PC (core i7, 3.4 GHz, no hyper threading, Ubuntu 12.04).
(Julia: 0.4.0-dev+6909).

The equivalent Python script shows an error of only 0.08 ms.
https://gist.github.com/ufechner7/1aa3e96a8a5972864cec

Not only the accuracy should be increased, but also the resolution. This is for example useful for rapid prototyping of control applications.

**Comment by stevengj** [comment]
Is 1.1 vs. 0.08 significant (in the statistical sense)?

**Comment by IainNZ** [comment]
What happens if you use `time_ns()`?

**Comment by stevengj** [comment]
Right, `time()` only has ms resolution.

**Comment by IainNZ** [comment]
```
IAINMAC:~ idunning$ python sleep.py
Average error is 0.0011s
IAINMAC:~ idunning$ julia sleep.jl
Average error is 0.002060507239999997 s
```

```julia
function check_sleep(amount)
    start = time_ns()
    sleep(amount)
    final = time_ns()
    delta = final - start
    return delta/(10.0^9)
end

total_error = zero(time_ns())
abs(check_sleep(0.050)-0.050)
for i in 1:100
    total_error += abs(check_sleep(0.050)-0.050)
end
println("Average error is $(total_error/100) s")
```

```python
from datetime import datetime
import time

def check_sleep(amount):
    start = datetime.now()
    time.sleep(amount)
    end = datetime.now()
    delta = end-start
    return delta.seconds + delta.microseconds/1000000.0

total_error = sum(abs(check_sleep(0.050)-0.050) for i in xrange(100))
print "Average error is %0.4fs" % (total_error/100)
```

**Comment by IainNZ** [comment]
In your test scripts you used a different sleep time, so your comparison isn't really accurate enough.
So less than a millisecond difference, could be due to timer accuracy or it could be overhead on sleep. I don't think `sleep` is really mean for `ms` scale stuff.

**Comment by ufechner7** [comment]
Using time_ns instead of time does not change the result. And the sleep() function of Julia has a resolution of only one ms. This is one thing, that should be changed. It is probably easy to fix. I will try to create a pull request tomorrow.

**Comment by ufechner7** [comment]
help?> time
search: time Timer time_ns timedwait mtime ctime @time @timev @timed DateTime

  time()

  Get the system time in seconds since the epoch, with fairly high (typically,
  microsecond) resolution.
So the problem is sleep(), not time(). The time() functions is implemented well.

**Comment by stevengj** [comment]
`time()` is ms resolution on Windows IIRC, but it is `gettimeofday` on Linux/Mac which is microseconds.

Isn't the `ms` resolution of `sleep` a libuv resolution?  `uv_timer_start` takes an [argument in milliseconds](http://docs.libuv.org/en/v1.x/timer.html).

**Comment by vtjnash** [comment]
https://www.kernel.org/doc/Documentation/timers/timers-howto.txt

tl;dr. don't sleep for <≈ 20ms (50 FPS) or the kernel will hate you (significantly lower throughput and higher power consumption, if it even decides to honor your request)

so this is largely by design, since it lets libuv use the cheaper `CLOCK_MONOTONIC_COARSE`

see also https://en.wikipedia.org/wiki/Timer_coalescing for various links covering other common platforms

(ps. if you've configured your own RTOS kernel to use HR clock hardware to control some device then you can ignore all of the above, but those devices usually aren't running Ubuntu)

**Comment by ufechner7** [comment]
Well, I do not agree with your statement "don't sleep for <= 20ms". If I have a real time motor controller and test it against a dynamic model, implemented in Julia with an update rate of 100 Hz, then my model needs 2.0 .. 8.9 ms for one simulation step and I need a delay of 1.1 .. 8.0 ms to wait, until the next time slice begins. This works well. If the accuracy of the sleep function would be improved, I spend 2.0 to 9.9 ms for one simulation step. I could do 10% more calculations without violating the real-time requirements. I would not care, "if the kernel hates me".

**Comment by lobingera** [comment]
@ufechner7, it's not his statement 'don't sleep for <= 20ms', it's rather like: You cannot force the task scheduler in a multitasking system do impossible things (this here is a longer story, but still).
The discussion about accuracy of sleep on unix kernels is a rather old one (just to show off: https://www.heise.de/artikel-archiv/ix/1994/11/168_Zeitspiel, and in the article we come to the conclusion, that without realtime support in the kernel...) and since then some people understood, that sleep(x) means to hand over the control to multitasking for at least but not exactly X time. The julia situation seems to be more pronounced, as inbetween your program and the kernel, there is the libuv.

I don't have a clear recommendation for you how to solve realtime requirements with julia running on a std. linux kernel, but you should lower your expections. Realtime on (any) unix system without kernel support is stretching the limitations (of time slice scheduling).

**Comment by ufechner7** [comment]
Well, I can only say that Python is currently doing 11 times better then
Julia. A low latency kernel with some tweaks, like no hyperthreading and
scheduling set to performance is already doing pretty well.

2015-08-24 12:28 GMT+02:00 Andreas Lobinger <notifications@github.com>:

**Comment by lobingera** [comment]
Python is not doing better, it just is running without another layer between running code and asking the kernel for resources.

**Comment by eschnett** [comment]
It seems there's a bit of reluctance to improve Julia's sleep,which is
something that (1) Uwe needs and (2) apparently already present in Python.
It should be straightforward to call `usleep` in Linux directly to
circumvent Julia:

usleep(usecs) = ccall(:usleep, Cint, (Cuint,), usecs)

-erik

On Mon, Aug 24, 2015 at 9:45 AM, Andreas Lobinger <notifications@github.com>
wrote:

> Python is not doing better, it just is running without another layer
> between running code and asking the kernel for resources.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/issues/12770#issuecomment-134205983>.
>



-- 
Erik Schnetter <schnetter@gmail.com>
http://www.perimeterinstitute.ca/personal/eschnetter/

**Comment by vtjnash** [comment]
> Well, I can only say that Python is currently doing 11 times better then Julia. A low latency kernel with some tweaks, like no hyperthreading and scheduling set to performance is already doing pretty well.

The general advice on the web is to just busy wait when dealing with such short time periods (http://www.geisswerks.com/ryan/FAQS/timing.html). Python is not doing "better" imo. This example works in your python code example because your system is unloaded and you don't care whether the kernel is executing a blocking busy wait.

In julia, you can access this same function by directly ccall'ing `nanosleep`, or the julia function wrapper `Libc.systemsleep` that exposes this behavior. Note that on windows, the specified timeout is rounded down to the nearest 10 ms tick (although this is configurable in userspace, per-process, and can be reduced to a 1 ms minimum interval).

> A low latency kernel with some tweaks, like no hyperthreading and scheduling set to performance...

this is exactly my argument for not changing this for the average user: it isn't supported by either the stock software or the average hardware. Note that 1 GHz / 0.1 ms is 100,000 cycles, so the kernel usually tries to avoid causing that much overhead. My (stock) ubuntu kernel for example is configured with `CONFIG_HZ=250` (looking at `/boot/config-3.13.0-40-generic`), which means the minimum preemption time slice is 4ms.

> then my model needs 2.0 .. 8.9 ms for one simulation step and I need a delay of 1.1 .. 8.0 ms to wait

or you could setup a repeating timer with a 10ms interval. and then as long as you never miss a deadline, you should get an average of 100 events / second.

**Comment by ufechner7** [comment]
I agree, that with a repeating timer you might not need sleep() in curtain use cases. I know, how to implement this in Python, but don't know, how to implement it in Julia yet. The Python version uses threads, that are not supported by Julia yet.
You write:
"My (stock) ubuntu kernel for example is configured with CONFIG_HZ=250 (looking at /boot/config-3.13.0-40-generic), which means the minimum preemption time slice is 4ms."
I always use a tickles kernel. See: http://www.quora.com/What-is-a-tickless-kernel . Better performance, no overhead.

From my current point of view I might call:
usleep_range(unsigned long min, unsigned long max)
from Julia as ccall and allow a jitter of 100us as suggested in https://www.kernel.org/doc/Documentation/timers/timers-howto.txt . Thank's for the link.

And I DO care if the kernel is doing a blocking, busy wait. This is not what I want, because in my simulation environment I usually have at least six processes, but only four CPU cores.

**Comment by vtjnash** [comment]
> I always use a tickles kernel.

my choice of terminology was inaccurate. I should have said "timeslice", since the two concepts are related, but not exactly the same.

> Better performance, no overhead.

neither of those assertions are universally true. running with a tickless kernel and an interrupt frequency less than the timeslice defeats the purpose of this feature (which is to reduce power consumption) and can cause performance issues on real-time problems due to the greater overhead of using a tickless kernel: https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/Documentation/timers/NO_HZ.txt#n86

> in my simulation environment I usually have at least six processes, but only four CPU cores.

this makes it sound like you are over-scheduling your CPU and using suboptimal queuing. typically you want your thread pool to have exactly the same number of worker threads as CPU cores, and to keep them 100% busy (including by doing busy waiting if they have an brief pause).

**Comment by StefanKarpinski** [comment]
@ufechner7, it seems like you are running a fairly special kernel, on special hardware and trying to solve a fairly special problem. Since Julia lets you make arbitrary C calls, it seems like you ought to be able to do what you need to here. At the same time, it doesn't seem like this change would be sensible for typical users of the language.

**Comment by ufechner7** [comment]
Well, I am researcher, which is probably a very typical user of Julia. I work at the Faculty of Aerospace Engineering of TU Delft. We switched from Matlab to Python in education 4 years ago. I would like to start to introduce Julia in education and research. But I can do this only if Julia is better then Python. Real time flight simulations are a typical usage of software at our faculty. Just my 2 cents.

**Comment by ufechner7** [comment]
And: "Be greedy. Want the best!" is the motto of Julia, isn't it!

**Comment by vtjnash** [comment]
but the asterisk on that says that Julia will not prevent you from writing bad code, but it may penalize it. just like the kernel is letting you run your tickless kernel, but may be penalizing you for it in performance and power consumption.

**Comment by StefanKarpinski** [comment]
@ufechner7, is there some feature missing that prevents you from doing what you need to here? It seems that you should be using `Libc.systemsleep` instead of `sleep()`, or better still, busy waiting.

**Comment by ufechner7** [comment]
Well, I did not try the alternatives to sleep() yet (and there are many). Busy waiting is not a good idea for my application, as I explained before. There is no need to come to a quick conclusion about the question, if Base.sleep should be improved or not. I think it should, to make live easy for beginners.

**Comment by vtjnash** [comment]
conditionally calling `Libc.systemsleep` if the user requests a timeout below 1ms seems like a disservice to beginners to me since you miss out on the discussion about why your program is not appropriately structured to avoid unnecessary kernel overhead.

**Comment by ufechner7** [comment]
So you would say, the behaviour of Python.sleep on Windows is better the the behaviour of Linux:
http://i.stack.imgur.com/GeF8u.png
Full thread:
http://stackoverflow.com/questions/1133857/how-accurate-is-pythons-time-sleep

**Comment by vtjnash** [comment]
it's more efficient and it's more honest about the actual real-world expected behavior, so yes. benchmarking an unloaded system doing a busy wait loop (or a lot of nothing) is practically useless.

**Comment by ufechner7** [comment]
How do you come to the conclusion that I was:
- benchmarking an unloaded system
- using a busy wait loop
You just make wrong assumptions without giving any evidence, that your assumptions are true.

**Comment by eschnett** [comment]
@vtjnash Instead of waiting for people to discover the problem with `sleep` and then explaining things in a discussion thread, maybe this should be mentioned in the documentation, along with work-arounds for those who know why they want to break the rules?

**Comment by StefanKarpinski** [comment]
It seems that there are a few issues here:

1. making `sleep` return closer to the requested time when possible
2. the fact that there are many circumstances where it's not possible

The second fact doesn't necessarily imply that we should not try to do the first thing, but it does make it more questionable. Python seems to have gone to great lengths to make sleep "accurate" when possible (single-threaded, unloaded system, real-time kernel). Julia uses coroutines for I/O and is likely in the near future to have true multithreading, all of which makes it harder to ever guarantee that sleep will be "accurate", so it's unclear if we should actually try to do what's not actually possible – something that almost works is worse than something that's more honest about it being impossible.

**Comment by lobingera** [comment]
@StefanKarpinski, where would i find more information about *is likely in the near future to have true multithreading* ?

**Comment by timholy** [comment]
https://github.com/JuliaLang/julia/commits/threading is the best resource I know of.

**Comment by StefanKarpinski** [comment]
Yes, that branch is the current state of work. There is ongoing work to merge this threading support back into master Julia so that you can enable threading with a compiler flag, with the longer term intention to have it enabled by default with a nice high-level user interface.

**Comment by josefsachsconning** [comment]
Just for the record, the threading support was merged into master at 8307c6d on Oct 30, correct?  Is there a new best resource for staying abreast of threading developments?

**Comment by tkelman** [comment]
@josefsachsconning probably the "multi-threading" issue/PR label: https://github.com/JuliaLang/julia/issues?q=sort%3Aupdated-desc+label%3Amulti-threading

**Comment by ufechner7** [comment]
There is now a package, that provides the timing performance, that I am locking for, at least for Linux:
https://github.com/ibadr/POSIXClock.jl

**Comment by stevengj** [comment]
Since `usleep` is was deprecated and removed from the POSIX spec, it might be a good idea anyway to have `Libc.systemsleep` call `nanosleep` or `clock_nanosleep` where possible.
